File: frontend/src/lib/kpi.ts
import { apiFetch } from "@/lib/api";
import type { KpiMetric, KpiMetricCreate, KpiDashboard, KpiBonusSummary } from "@/types";

export async function createOrUpdateMetric(data: KpiMetricCreate, isServer = false): Promise<KpiMetric> {
  return apiFetch<KpiMetric>("/kpi/", { method: "POST", body: JSON.stringify(data), isServer });
}

export async function getKpiMetrics(month?: number, year?: number, isServer = false): Promise<KpiMetric[]> {
  const params = new URLSearchParams();
  if (month) params.set("month", String(month));
  if (year) params.set("year", String(year));
  const qs = params.toString();
  return apiFetch<KpiMetric[]>(`/kpi/${qs ? `?${qs}` : ""}`, { isServer });
}

export async function getKpiDashboard(month: number, year: number, isServer = false): Promise<KpiDashboard> {
  return apiFetch<KpiDashboard>(`/kpi/dashboard?month=${month}&year=${year}`, { isServer });
}

export async function getBonusSummary(month: number, year: number, isServer = false): Promise<KpiBonusSummary[]> {
  return apiFetch<KpiBonusSummary[]>(`/kpi/bonus-summary?month=${month}&year=${year}`, { isServer });
}

File: backend/app/api/v1/endpoints/kpi.py
import uuid
from typing import List, Optional
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.api.dependencies import get_current_user, require_admin
from app.models import User
from app.services.kpi_service import KpiService
from app.schemas.kpi import KpiMetricCreate, KpiMetricResponse, KpiDashboard, KpiBonusSummary

router = APIRouter()


@router.post("/", response_model=KpiMetricResponse, status_code=status.HTTP_201_CREATED, summary="Criar/Atualizar Métrica KPI")
def create_or_update_metric(
    data: KpiMetricCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    service = KpiService(db=db, tenant_id=current_user.tenant_id)
    return service.create_or_update_metric(data=data)


@router.get("/", response_model=List[KpiMetricResponse], summary="Listar Métricas KPI")
def list_metrics(
    month: Optional[int] = None,
    year: Optional[int] = None,
    skip: int = 0,
    limit: int = 50,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = KpiService(db=db, tenant_id=current_user.tenant_id)
    return service.list_metrics(month=month, year=year, skip=skip, limit=limit)


@router.get("/dashboard", response_model=KpiDashboard, summary="Dashboard KPI")
def get_kpi_dashboard(
    month: int,
    year: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = KpiService(db=db, tenant_id=current_user.tenant_id)
    return service.get_dashboard(month=month, year=year)


@router.get("/bonus-summary", response_model=List[KpiBonusSummary], summary="Resumo de Bonificação P4P")
def get_bonus_summary(
    month: int,
    year: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = KpiService(db=db, tenant_id=current_user.tenant_id)
    return service.get_bonus_summary(month=month, year=year)


@router.get("/{metric_id}", response_model=KpiMetricResponse, summary="Detalhe da Métrica KPI")
def get_metric(
    metric_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = KpiService(db=db, tenant_id=current_user.tenant_id)
    return service.get_metric(metric_id=metric_id)
